oh/conversation/conversation_abc.py

```diff
@@ -0,0 +1,149 @@
+    ) -> UUID:
+        """Add a listener and return a unique id for it"""
+
+    @abstractmethod
+    async def remove_listener(self, listener_id: UUID) -> bool:
+        """Remove the listener with the id given"""
+
+    @abstractmethod
+    async def trigger_event(self, detail: EventDetailABC) -> oh_event.OhEvent:
+        """Trigger the event given"""
+
+    @abstractmethod
+    async def get_event(self, event_id: UUID) -> Optional[oh_event.OhEvent]:
+        """Get an event with the id given"""
+
+    @abstractmethod
+    async def search_events(
+        self, filter: Optional[EventFilter] = None, page_id: Optional[str] = None
+    ) -> Page[oh_event.OhEvent]:
+        """Search events in this conversation"""
+
+    @abstractmethod
+    async def count_events(self, filter: Optional[EventFilter] = None) -> int:
+        """Count events in this conversation"""
+
+    @abstractmethod
+    async def create_task(
+        self,
+        runnable: runnable_abc.RunnableABC,
+        title: Optional[str] = None,
+        delay: float = 0,
+    ) -> oh_task.OhTask:
+        """Create a task and return details of it.  Throw a SesionError if the conversation is not in a ready state. """
+
+    @abstractmethod
+    async def run_task(
+        self,
+        conversation_id: UUID,
+        runnable: runnable_abc.RunnableABC,
+        title: Optional[str] = None,
+        delay: float = 0,
+    ):
+        """Run the task and return the result. Throw a SesionError if the conversation is not in a ready state. """
+
+    @abstractmethod
+    async def cancel_task(self, task_id: UUID) -> bool:
+        """Attempt to cancel task with the id given. Return true if the task was cancelled, false otherwise"""
+
+    @abstractmethod
+    async def get_task(self, task_id: UUID) -> Optional[oh_task.OhTask]:
+        """Get the handle for the task with the id given"""
+
+    @abstractmethod
+    async def search_tasks(
+        self, filter: TaskFilter, page_id: str
+    ) -> Page[oh_task.OhTask]:
+        """Get info on running tasks"""
+
+    @abstractmethod
+    async def count_tasks(self, filter: TaskFilter) -> int:
+        """Get the number of tasks"""
+
+    @abstractmethod
+    async def create_dir(self, path: str) -> FileInfo:
+        """
+        Make the directory at the path given if it does not exist. Return info in the directory.
+        Directories have the mime type `application/x-directory`
+        """
+
+    @abstractmethod
+    async def create_file(self, path: str) -> FileInfo:
+        """Update the updated_at on the file at the path given. If no file exists, create an empty file"""
+
+    @abstractmethod
+    async def save_file(self, path: str, content: Union[Path, IOBase]) -> FileInfo:
+        """Upload a file to the path given. If no file exists, create an empty file. If a file exists, overwrite it."""
+
+    @abstractmethod
+    async def delete_file(self, path: str) -> bool:
```

Comment:
Looks like the Conversation is tightly coupled to the runtime, which makes it difficult to e.g. swap in different runtimes without re-implementing the whole Conversation interface

Currently the Session gets an abstract `Runtime` object, which I think will work better when we have multiple runtimes available.